src/c-writer.cc

```diff
@@ -2300,15 +2425,42 @@ void CWriter::Write(const Func& func) {
     Write(CloseBrace(), Newline());
   }
 
-  stream_ = c_stream_;
+  /*
+   * Choose the stream that this function goes in, based on its name.
+   * We need a hash function that's consistent across runs of
+   * wasm2c (which std::hash doesn't guarantee) because if only one function
+   * changes across runs, or one function is added or removed, we want only one
+   * stream to change and need to be recompiled. Here we use "SHA-256/64"
+   * because it's sufficient and already available in wasm2c.
+   */
+  std::string hashed_function_name;
+  sha256(func.name, hashed_function_name);
```

Comment:
Why not simply sort the function by name and divide the results in on N equal sized buckets?   It would result in the nice property that each of the N files would have have exactly M/N functions in it.  It seems like a more obvious and straight forward approach.

It would still be the case the changing just one function would result in only one file changing.   We would loose the property when adding a removing functions.. but I don't think that sounds like a particularly important case (I could be wrong?).

I'd also rather not increase our dependency the external sha256 codebase simply because we've already pulled it in.

Reply:
It is indeed the case that for the majority of users of wasm2c, sorting functions by name and dividing the results into equal buckets is the more straight forward approach. However, in our use case of wasm2c, we expect users to develop their programs in an incremental way, and therefore remaining the function distribution stable when a single function is added or removed is important to us.

Would it be better to add a field to `WriteCOptions` where callers can specify their own functions for deciding how functions are distributed across buckets, and the default behavior is dividing functions based on function name sorting?

Reply:
Have you measured this to have a real effect on the programs you are working on?   Are you saying you have multiple large wasm programs that you rebuild frequently, with only difference being that addition of removal of a single function?   What kind of large programs are we talking about here?  How long are you seeing the compile times?  Are you using some kind of build system that doesn't use timestamps for change detection?

Reply:
In our use case, we take a single WebAssembly module as input from the user, using wasm2c+clang to transform that to x86 for execution.

We wrote a program that contains an entry function, some helper functions. The entry function invokes the backend of Clang/LLVM for compiling c source code into x86. We compiled this program to a WebAssembly module. wasm2c + clang (c to x86) on this module would take ~2 hours without any optimization flag and ~7 hours with "-O2".

In this case, we would hope that adding or removing functions from the part we wrote ourselves can be done by only recompiling a small number of .c outputs.